Add notes and questions for n17331 (#335)
bitcoin-core-review-club/website: _posts/2021-04-28-#17331.md

---
layout: pr
date: 2021-04-28
title: "Use effective values throughout coin selection"
pr: 17331
authors: [achow101]
components: ["wallet"]
host: Xekyo
status: upcoming
commit: 4ac1add
---

## Notes

- Wallets maintain their balance as a pool of Unspent Transaction Outputs
  (UTXOs).

- _Coin selection_ refers to the process of picking UTXOs from the wallet's UTXO
  pool to fund a transaction.

- Beyond the primary goal of funding a transaction, the secondary goals of coin
  selection are:

    - Minimizing short term and long term fees:  
      We want to both pay the least right now, but we also want to be good
      stewards of our UTXO pool and not rack up large UTXO counts that will

>       translate to huge future transaction fees.

Reviewer: Doesn't totally fit the "short term and long term fees" description. I would think that it's more about a user wanting to pay the minimum amount to get their tx confirmed in time. Including long term fees e.g. spending change in the future

Author: I've rewritten this section to:

>    - Minimizing short term and long term fees:  
>       We have to balance two needs, we want to pay the smallest fee now to get a
>       timely confirmation, but we have to keep in mind that all of our UTXOs
>       need to be spent in the future. So, we shouldn't over-optimize locally at
>       the detriment of large future costs. E.g. always using largest-first
>       selection minimizes the current cost, but grinds your UTXO pool to dust.

Does that make more sense?

Reviewer: yes!!!

    - Maintaining financial privacy:  

>       Ideally, our transactions leak as little information about ourselves and

Reviewer: This is a bit abstract, maybe add a little about linking addresses, e.g. spending UTXOs together usually means it's from the same wallet.

Author: Rewrote the section

      our business partners as possible.

    - Reliably move the payment(s) to finalization:  
      Reduce the chance of transactions being conflicted and transfer value in a

>       limited time-frame

Reviewer: Again a little too abstract and unclear how it relates to coin selection. Maybe bring the description below here!

Author: Done!

- Using unconfirmed inputs can make transactions unreliable. When a foreign
  input is used, the parent transaction may time out or get replaced. Even using
  self-sent unconfirmed funds may delay the new transaction if the parent
  transaction has an extensive ancestry, is extraordinarily large, or used a

>   lower feerate than targeted for the child transaction.

Reviewer: This could be moved to the "Reliably move the payments to finalization" section above!

Author: Great idea, done!

> - When a new transaction is built, the target feerate is provided as an input.

Reviewer: I think a more helpful note would be a rough outline of the process of creating a tx, with a bit of extra detail for coin selection specifically:

1. `CreateTransaction()` builds the header and payment outputs (no change yet), which are fixed.
2. Grab feerates using `GetMinimumFeeRate()` (goes through fee estimation and/or user configurations). All of these are fixed: the tx header, payment outputs, target feerate, target amount, etc.
3. Get all `AvailableCoins()`
4. Use `SelectCoins()` to get the inputs (fund the transaction). 
  (4a) This uses pre-selected coins first (`CoinControl`), then groups UTXOs into `OutputGroup`s, and runs `SelectCoinsMinConf()` with increasingly permissive `CoinEligibilityFilter`s until we find a set of coins.
   (4b) Each iteration of `SelectCoinsMinConf()` tries BnB, then Knapsack if that fails.
5. Sign the transaction

Author: Thanks, this was super helpful. I've added it with some slight modifications.

  which in turn increases the necessary fees to maintain the targeted feerate.
  We can preempt the cost added by the input by considering each UTXO at its
  _effective value_, its value reduced by the UTXO's input cost at the given
  feerate.

- Using _effective values_ in coin selection means that we can freely add and
  remove inputs until we have have raised enough funds to pay for the
  transaction's fixed costs: the payment amounts, the transaction header and the
  output data. Whatever final input set we settle for, the increased transaction
  size and input costs have already been accounted for.

- The _knapsack solver_ sorts the available UTXOs by effective value, and then
  combines candidate input sets by randomly selecting or skipping inputs until
  it has sufficient funds for the transaction. When it finds a viable solution,
  it keeps the solution if it produces a smaller change than the prior stored
  solution. Then, the algorithm deselects the last input and continues
  traversing the UTXO list as described, trying to converge on the smallest
  change output above `MIN_CHANGE`. This random process is repeated up to 1,000
  times.

> - The _Branch and Bound solver (BnB)_ deterministically searches the complete

Reviewer: I think you should mention the no-change part in this first sentence, because it's a crucial poitn in this PR.

Author: Good point, moved it up.

  combination space. It performs a depth-first search on a binary tree where
  each branch represents the inclusion or exclusion of a UTXO. It explores each
  inclusion branch first, and backtracks whenever a subtree cannot yield a
  solution. It returns the first discovered input set that is an "exact match"
  for the funding requirement of the transaction. A resulting transaction avoids
  the creation of a change output, but the BnB algorithm is not guaranteed to
  find a solution even when there are sufficient funds.

- An
  [OutputGroup](https://github.com/bitcoin-core-review-club/bitcoin/blob/4ac1adda9914d845aaea5804af4801ffec53c701/src/wallet/coinselection.h#L72)
  collects all UTXOs that were sent to the same public key.

> - This PR implements the use of _effective values_ across all different coin

Reviewer: Also mention that the PR makes `SelectCoinsMinConf` try both strategies (i.e. BnB then Knapsack if needed) instead of taking a bool.

Author: Added a bullet point!

  selection strategies. Prior, it was only used for the BnB selection where it
  facilitated the search to be effectively performed. Introducing _effective
  values_ into the knapsack solver allows some of the pre-selection steps to be
  shared between BnB and knapsack, and prevents that the knapsack solver needs
  to be repeated when the resulting input set exceeds the previously allocated
  fee budget.

## Questions

> 1. How does the _effective value_ differ from the `m_value` of a UTXO?

Reviewer: I think I'd actually change this to `nValue` because that's what output amounts are usually called in specs

Author: Done!

2. What is `not_input_fees`? What does this help with?

3. What happens when `m_subtract_fee_outputs` is true?

4. What does `cost_of_change` represent?

> 5. How do we know the `cost_of_change` in advance?

Reviewer: These can be combined into 1 question

Author: Done

6. What are `discard_feerate`, `long_term_feerate`, and `effective_feerate`?

7. Why do the `OutputGroups` get calculated separately for BnB and Knapsack?

> 8. What purpose did the while-loop in `CreateTransaction()` serve that gets removed?

Reviewer: All of these need links to the code

Author: Gonna push my current iteration, then add links.

       then used in one or multiple rounds of `SelectCoinsMinConf()` which
       use increasingly permissive `CoinEligibilityFilter`s to preselect
       `OutputGroup`s until an input set is found. As described above,
       `SelectCoinsMinConf()` tries BnB first, then falls back to knapsack if no
       solution was found.
    5. Determine whether to create a change output, finish assembling the
       transaction, and sign the transaction.

- When a UTXO is selected as an input, the size of the transaction is increased
  which in turn increases the necessary fees to maintain the targeted feerate.
  We can preempt the cost added by the input by considering each UTXO at its
  _effective value_, its value reduced by the UTXO's input cost at the given
  feerate.

- Using _effective values_ in coin selection means that we can freely add and
  remove inputs until we have have raised enough funds to pay for the
  transaction's fixed costs: the payment amounts, the transaction header and the
  output data. Whatever final input set we settle for, the increased transaction
  size and input costs have already been accounted for.

> - The _knapsack solver_ sorts the available UTXOs by effective value, and then

Reviewer: Ordering: I think these could be moved up, i.e.
```
- There are 2 types of solvers
    - knapsack
    - branch and bound
- When a UTXO is selected... effective values!

```

Author: Reordered.

  it has sufficient funds for the transaction. When it finds a viable solution,
  it keeps the solution if it produces a smaller change than the prior stored
  solution. Then, the algorithm deselects the last input and continues
  traversing the UTXO list as described, trying to converge on the smallest
  change output above `MIN_CHANGE`. This random process is repeated up to 1,000
  times.

- The _Branch and Bound solver (BnB)_ deterministically searches the complete
  combination space to find an input set that will avoid the creation of a
  change output. It performs a depth-first search on a binary tree where each
  branch represents the inclusion or exclusion of a UTXO, exploring inclusion
  branches first, and backtracking whenever a subtree cannot yield a solution.
  It returns the first discovered input set that is an "exact match" for the
  funding requirement of the transaction. The BnB algorithm is not guaranteed to
  find a solution even when there are sufficient funds since an exact match may
  not exist.

- An
  [OutputGroup](https://github.com/bitcoin-core-review-club/bitcoin/blob/4ac1adda9914d845aaea5804af4801ffec53c701/src/wallet/coinselection.h#L72)

>   collects all UTXOs that were sent to the same public key.

Reviewer: I think this can either be deleted or moved up to the first bullet

Author: Moved up.

  facilitated the search to be effectively performed. Introducing _effective
  values_ into the knapsack solver allows some of the pre-selection steps to be
  shared between BnB and knapsack, and prevents that the knapsack solver needs
  to be repeated when the resulting input set exceeds the previously allocated
  fee budget.

- This PR also changes the behavior of `SelectCoinsMinConf()` to try BnB first
  and then attempt knapsack only in the case that BnB did not find an input set.
  Before this PR, the function would only run one or the other strategy
  controlled by the boolean flag `use_bnb`.

## Questions

1. How does the _effective value_ differ from the `nValue` of a UTXO?

2. What is `not_input_fees`? What does this help with?

3. What happens when `m_subtract_fee_outputs` is true?

4. What does `cost_of_change` represent? How do we know the `cost_of_change` in advance?

5. What are `discard_feerate`, `long_term_feerate`, and `effective_feerate`?

6. Why do the `OutputGroups` get calculated separately for BnB and Knapsack?

7. What purpose did the while-loop in `CreateTransaction()` serve that gets removed?

> 8. Under which circumstances do resulting transactions not create a change output?

Reviewer: ```suggestion
8. Under which circumstances do resulting transactions not create a change output? (Hint: there are two cases [here](https://github.com/bitcoin/bitcoin/blob/4ac1adda9914d845aaea5804af4801ffec53c701/src/wallet/wallet.cpp#L2919))
```

Author: I like the "Hint" style. Took that

---
layout: pr
date: 2021-04-28
title: "Use effective values throughout coin selection"
pr: 17331
authors: [achow101]
components: ["wallet"]
host: Xekyo
status: upcoming
commit: 4ac1add
---

## Notes

- Wallets maintain their balance as a pool of Unspent Transaction Outputs
  (UTXOs). An
  [OutputGroup](https://github.com/bitcoin-core-review-club/bitcoin/blob/4ac1adda9914d845aaea5804af4801ffec53c701/src/wallet/coinselection.h#L72)
  collects all UTXOs that were sent to the same public key.

- _Coin selection_ refers to the process of picking UTXOs from the wallet's UTXO
  pool to fund a transaction.

- Beyond the primary goal of funding a transaction, the secondary goals of coin
  selection are:

    - Minimizing short term and long term fees:  
      We have to balance two needs, we want to pay the smallest fee now to get a
      timely confirmation, but we have to keep in mind that all of our UTXOs
      need to be spent in the future. So, we shouldn't over-optimize locally at
      the detriment of large future costs. E.g. always using largest-first

>       selection minimizes the current cost, but grinds your UTXO pool to dust.

Author: The suggestion does not capture what I intended to express. I've amended the sentence, hopefully it's clearer.

    - Maintaining financial privacy:  
      There are a number of heuristics that tracking companies employ to cluster
      payments and addresses. For example, using inputs with the same output
      script in two transactions indicates that the two transactions involved
      the same party. Similarly, it is usually assumed that all inputs of a
      transaction were controlled by the same entity. Sometimes the privacy
      and economic considerations are opposed. If you want to understand privacy
      considerations better, check out the [Privacy
      article](https://en.bitcoin.it/wiki/Privacy) on the Bitcoin wiki.

    - Reliably move the payment(s) to finalization:  

>       Using unconfirmed inputs can make transactions unreliable. When a foreign
>       input is used, the parent transaction may time out or get replaced. Even

Reviewer: I know they're the same thing in this context but inputs could be confusing
```suggestion
      Using unconfirmed outputs can make transactions unreliable. Unconfirmed
      transactions received from another wallet may time out
      or be replaced, making those funds disappear. Even
```

Author: done

      using self-sent unconfirmed funds may delay the new transaction if the
      parent transaction has an extensive ancestry, is extraordinarily large, or
      used a lower feerate than targeted for the child transaction.

- The process of creating a transaction in Bitcoin Core's wallet roughly
  follows these steps:

    1. `CreateTransaction()` builds the header and recipient outputs, i.e. the
       parts of the transaction that will remain fixed throughout the coin
       selection process.

>     2. The feerates are established using `GetMinimumFeeRate()` which falls back

Reviewer: I think you might want to quickly mention which feerates, otherwise it might not be obvious why there are multiple

Author: Mentioned some feerates

       either to fee estimation or user input. The feerates remain constant for
       the remaining process. The amount to be raised from the inputs,
       `nTargetValue` is calculated from the sum of recipient amounts, the
       target feerate, and the size of the fixed transaction parts.

    3. The wallet's UTXO pool is retrieved via `AvailableCoins()`.

    4. Using the above fixed values, `SelectCoins()` is used to pick an input
       set to fund the transaction. If the user manually selected some UTXOs by
       means of the `CoinControl` feature, these UTXOs are used first. The
       remaining UTXOs are grouped into `OutputGroup`s. The grouped UTXOs are

>        then used in one or multiple rounds of `SelectCoinsMinConf()` which
>        use increasingly permissive `CoinEligibilityFilter`s to preselect

Reviewer: Otherwise I think it sounds like `SelectCoinsMinConf()` is doing it

```suggestion
       then used in one or multiple rounds of `SelectCoinsMinConf()`, using
       increasingly permissive `CoinEligibilityFilter`s to preselect
```

Author: Done

       `OutputGroup`s until an input set is found.

    5. Determine whether to create a change output, finish assembling the
       transaction, and sign the transaction.

- When a UTXO is selected as an input, the size of the transaction is increased
  which in turn increases the necessary fees to maintain the targeted feerate.
  We can preempt the cost added by the input by considering each UTXO at its
  _effective value_, its value reduced by the UTXO's input cost at the given

>   feerate.

Reviewer: Move this below the solvers bullet

Author: The solver explanations make use of `effective_values` in their explanation. Moving this would make the definition of that follow the use of it.

Reviewer: ah true okay

    - The _Branch and Bound solver (BnB)_ deterministically searches the
      complete combination space to find an input set that will avoid the
      creation of a change output. It performs a depth-first search on a binary
      tree where each branch represents the inclusion or exclusion of a UTXO,
      exploring inclusion branches first, and backtracking whenever a subtree
      cannot yield a solution.  It returns the first discovered input set that
      is an _exact match_ for the funding requirement of the transaction. The
      qualifier _"exact match"_ here refers here to an input set that
      overshoots the `nTargetValue` by less than the cost of a change output.
      The BnB algorithm is not guaranteed to find a solution even when there
      are sufficient funds since an _exact match_ may not exist.

- Using _effective values_ in coin selection means that we can freely add and
  remove inputs until we have have raised enough funds to pay for the
  transaction's fixed costs: the payment amounts, the transaction header and the
  output data. Whatever final input set we settle for, the increased transaction
  size and input costs have already been accounted for.

> - This PR implements the use of _effective values_ across all different coin

Reviewer: ```suggestion
- [PR #17331](https://github.com/bitcoin/bitcoin/pull/17331) implements the use of _effective values_ across all different coin
```

Author: Ack

  selection strategies. Prior, it was only used for the BnB selection where it

>   facilitated the search to be effectively performed. Introducing _effective
>   values_ into the knapsack solver allows some of the pre-selection steps to be
>   shared between BnB and knapsack, and prevents that the knapsack solver needs
>   to be repeated when the resulting input set exceeds the previously allocated
>   fee budget.

Reviewer: ```suggestion
  facilitated the search to be effectively performed. Introducing _effective
  values_ into the knapsack solver means it no longer works on a moving
  target, and thus only needs to be run once. It also allows some of the
  pre-selection steps to be shared between BnB and knapsack.
```

Author: Good, done

  Before this PR, the function would only run one or the other strategy
  controlled by the boolean flag `use_bnb`.

## Questions

1. How does the _effective value_
   [differ](https://github.com/bitcoin-core-review-club/bitcoin/commit/453c00c8eaffb7ee16b3b3232a6c3e39b1b52882#diff-1f2db0e4d5c12d109c7f0962333c245b49b696cb39ff432da048e9d6c08944d8R2473)
   from the `nValue` of a UTXO?

2. What is
   [`not_input_fees`](https://github.com/bitcoin-core-review-club/bitcoin/commit/453c00c8eaffb7ee16b3b3232a6c3e39b1b52882#diff-1f2db0e4d5c12d109c7f0962333c245b49b696cb39ff432da048e9d6c08944d8R2402)?
   What does this help with?

3. What happens when
   [`m_subtract_fee_outputs`](https://github.com/bitcoin-core-review-club/bitcoin/commit/453c00c8eaffb7ee16b3b3232a6c3e39b1b52882#diff-1f2db0e4d5c12d109c7f0962333c245b49b696cb39ff432da048e9d6c08944d8R2473)
   is true?

4. What does
   [`cost_of_change`](https://github.com/bitcoin-core-review-club/bitcoin/commit/453c00c8eaffb7ee16b3b3232a6c3e39b1b52882#diff-1f2db0e4d5c12d109c7f0962333c245b49b696cb39ff432da048e9d6c08944d8R2416)
   represent? How do we know the `cost_of_change` in advance?

5. What

>    [are](https://github.com/bitcoin-core-review-club/bitcoin/blob/4ac1adda9914d845aaea5804af4801ffec53c701/src/wallet/wallet.h#L611)

Reviewer: why is the link on "are" ?

Author: Moved link to feerates

   `long_term_feerate`, `effective_feerate`, and
   [`discard_feerate`](https://github.com/bitcoin-core-review-club/bitcoin/blob/4ac1adda9914d845aaea5804af4801ffec53c701/src/wallet/wallet.h#L69)?

> 6. Why do the `OutputGroups` get calculated separately for

Reviewer: ```suggestion
6. Why are the `OutputGroups` calculated separately for
```

Author: Done

   [BnB](https://github.com/bitcoin-core-review-club/bitcoin/blob/4ac1adda9914d845aaea5804af4801ffec53c701/src/wallet/wallet.cpp#L2415)
   and
   [knapsack](https://github.com/bitcoin-core-review-club/bitcoin/blob/4ac1adda9914d845aaea5804af4801ffec53c701/src/wallet/wallet.cpp#L2420)?

> 7. What purpose did the
>    [while-loop](https://github.com/bitcoin-core-review-club/bitcoin/commit/23c85ae91ea0a720b591cab8dfd20be72425ab31#diff-1f2db0e4d5c12d109c7f0962333c245b49b696cb39ff432da048e9d6c08944d8L2856)
>    in `CreateTransaction()` serve that gets removed?

Reviewer: ```suggestion
7. What purpose did the
   [while-loop](https://github.com/bitcoin-core-review-club/bitcoin/commit/23c85ae91ea0a720b591cab8dfd20be72425ab31#diff-1f2db0e4d5c12d109c7f0962333c245b49b696cb39ff432da048e9d6c08944d8L2856)
   in `CreateTransaction()` serve? Why is it safe to remove it?
```

Author: :+1:

---
layout: pr
date: 2021-04-28
title: "Use effective values throughout coin selection"
pr: 17331
authors: [achow101]
components: ["wallet"]
host: Xekyo
status: upcoming
commit: 4ac1add
---

## Notes

- Wallets maintain their balance as a pool of Unspent Transaction Outputs
  (UTXOs).

- _Coin selection_ refers to the process of picking UTXOs from the wallet's UTXO
  pool to fund a transaction.

> - What are the goals of coin selection? Economics, privacy, reliability of
>   inputs/conflicts

Reviewer 2: It's a bit confusing to form this as a question. Better would be a statement:

```suggestion
- The goals of coin selection are:
    - Minimizing fees
    - Maintaining privacy
    - Ensuring reliability of inputs/reducing the chance of transactions being conflicted.
```

Author: Filled out that section

> - Why are we trying to not spend unconfirmed inputs?

Reviewer 2: Should this be in the questions section?

Author: Added notes to address this question

- When a new transaction is built, the target feerate is provided as an input.

> - A UTXOs _effective value_ is its value reduced by its cost to be spent at a

Reviewer 2: ```suggestion
- A UTXO's _effective value_ is its value reduced by its cost to be spent at a
```

Author: D'oh fixed

>   given feerate.

Reviewer 2: Perhaps go into a bit more detail here. Why does an input have a 'cost to be spent at a given feerate'.

Author: Have elaborated this point

- Using _effective values_ in coin selection means that the cost of spending the
  UTXO has already been accounted for when a UTXO gets picked.

> - The knapsacksolver tries to combine input sets by randomly selecting or

Reviewer 2: ```suggestion
- The _knapsack solver_ tries to combine input sets by randomly selecting or
```

Author: Done, and extended the section.

  skipping inputs until it has sufficient funds for the transaction. When it
  finds a solution, it deselects the last input and tries to find one with a
  smaller change output. It repeats this random drawing 1000 times.

> - Branch and Bound deterministically searches the complete combination space. It

Reviewer 2: ```suggestion
- _Branch and Bound_ deterministically searches the complete combination space. It
```

Author: Done

  walks a binary search tree where each branch is the inclusion or exclusion of
  a UTXO. It explores the inclusion branch first, and backtracks whenever a
  subtree cannot yield a solution. It returns the first input set that is an
  exact match for the funding requirement of the transaction.

> - OutputGroups collect all UTXOs that were sent to the same public key.

Reviewer 2: ```suggestion
- An [OutputGroup](link to definition in code) collects all UTXOs that were sent to the same public key.
```

Author: Done

> 

Reviewer 2: Maybe add a bullet or two about what this PR is changing.

Author: Added a big bullet :)

## Questions

> 1. How do _effective values_ differ from the `m_value` of UTXOs?

Reviewer 2: ```suggestion
1. How does the _effective value_ differ from the `m_value` of a UTXO?
```

Author: Done

2. What is `not_input_fees`? What does this help with?

3. What happens when `m_subtract_fee_outputs` is true?

4. What does `cost_of_change` represent?

5. How do we know the `cost_of_change` in advance?

6. What are `discard_feerate`, `long_term_feerate`, and `effective_feerate`?

7. Why do the `OutputGroups` get calculated separately for BnB and Knapsack?

> 8. Why are `use_bnb` and `bnb_used` now obsolete?  // perhaps unnecessary

Reviewer 2: Maybe remove this

Author: Removed

> 9. What purpose did the while loop in CreateTransaction serve that gets removed?

Reviewer 2: ```suggestion
9. What purpose did the while loop in `CreateTransaction()` serve that gets removed?
```

Author: Is there a best practice how to link to the commit diff here in a future-proof way?

> 10. When do we not make change? // 2nd last commit

Reviewer 2: I don't understand this question/comment.

Author: Rewritten this question, hope it makes sense now!
